### ISSUE TITLE
Support Audio/Video compressed source

### ISSUE THREAD

**Issue by frozeus** [EricssonResearch/openwebrtc#224]
OpenWebRTC should be able to use compressed source H264/VP8 for video src and PCMA/PCMU/Opus for audio src and try to negociate to use this codec as default to avoid a recoding/resampling.

There is already a small discussion about #172

**Comment by ford-prefect** [comment]
This is sort of available statically via https://github.com/ford-prefect/openwebrtc/commits/compressed

Here's a rough task breakdown to make all this work dynamically as well.

* On the GStreamer side:
 * Fix up ```singledecodebin``` and merge -- https://bugzilla.gnome.org/show_bug.cgi?id=743511
 * My original code uses ```interapp*``` (https://bugzilla.gnome.org/show_bug.cgi?id=743510), but this would probably just end up being a subset of what we'd do for #240.

* On the OWR side:
  * Source reconfiguration: the source needs to be able to reconfigure itself if it sees new caps requested (new format, new height/width/framerate) parameters. The fact that we're using decodebin and encodebin downstream with the work I have already done should mean
that we can switch seamlessly (I have not tested this, however).
  * Test reconfiguration with whatever replaces interaudio/video.

**Comment by superdump** [comment]
Some compressed sources, such as rpicamsrc, support force key unit. If this is not yet proxied between the interapp* or whatever, I guess that needs to be supported. Otherwise you're basically forced to use periodic intra refresh.

**Comment by superdump** [comment]
After #240 is landed, we just need to deal with the conversion elements and proper signalling and configuration of profile, level etc where appropriate.

**Comment by suganthikarthick** [comment]
Hi,

We would need to take the H264 compressed video source as input in the openwebrtc stack.

In the following branch, which all commits I have to use for this functionality?
https://github.com/ford-prefect/openwebrtc/commits/compressed

This requires changes in gstreamer as well?

Thanks.

**Comment by suganthikarthick** [comment]
Hi Ford-perfect,

We tried to apply the gst patch for the interappsrc and the interappsink.
And the compilation also went fine.

But when we give the "gst-inspect-1.0 interappsrc", we are getting the following error.

(gst-plugin-scanner:24569): GStreamer-WARNING **: Failed to load plugin '/opt/openwebrtc-0.3/lib/gstreamer-1.0/libgstinter.so': /opt/openwebrtc-0.3/lib/gstreamer-1.0/libgstinter.so: undefined symbol: gst_inter_app_src_get_type
No such element or plugin 'interappsrc'

Do you have any suggestion on this?
Thanks.

**Comment by ford-prefect** [comment]
Hi @suganthikarthick, it looks like your build is missing something, but I'm not sure.

That said, you probably want to redo this branch based on @sdroege's work from #240. You can likely reuse quite a bit of what I did.

**Comment by suganthikarthick** [comment]
Thank you.
We have not yet started to modify the openwebrtc stack as per your changes.
We applied the patch for including the interappsrc and interappsink for gstreamer only.

After applied the patch, we compiled the entire gst-plugins-bad and tried to check with the "gst-inspect-1.0 interappsrc"
 But we are facing the above mentioned error.
Thanks.

**Comment by ford-prefect** [comment]
Did gstinterappsrc.c get correctly compiled and linked into your plugin?

**Comment by suganthikarthick** [comment]
Hi,
It is working fine now.. I missed to give a "configure" and then build.

Thanks.

**Comment by suganthikarthick** [comment]
Hi,
We have applied the patch to both openwebrtc and the gst-plugins-bad.
We have a general question that whether it is possible to use "custom gst" for video input in the openwebrtc application?

Thanks.

**Comment by suganthikarthick** [comment]
Hi,

When we are using the local video source instead of custom gst,  we are getting the following errors.

Error in element video-source.
Error: Internal data flow error.
Debugging info: gstbasesrc.c(2943): gst_base_src_loop (): /GstPipeline:local-video-capture-source-bin-0/GstV4l2Src:video-source:
streaming task paused, reason not-negotiated (-4)

Thanks.

**Comment by suganthikarthick** [comment]
Hi Ford-perfect,

Now we have applied the patch for both gst-plugins-bad and for the openwebrtc stack and compiled and installed the libraries.

When running the application, we are getting the following errors.

0:00:10.393169524 12380 0xdcc290 ERROR owrtransportagent owr_transport_agent.c:776: handle_new_send_source: Failed to link "(null)" with transport bin

We tried using both custom-gst and with local video source as well. The video is a H264 compressed input. 
It would of great help if somebody help us in resolving the issue.

Thank you.

Thanks.

**Comment by ford-prefect** [comment]
Unfortunately I don't have the bandwidth to look at this right now. That said, I'd first go over each of my patches to validate that they are still applicable on the current OWR. The new owr inter elements should make my interapp work obsolete, and you might need to modify my patches a bit to work with master. The basic approach should be no different from what I did, it's mostly a matter of making sure the new inter elements are hooked up correctly to deal with compressed formats.

**Comment by suganthikarthick** [comment]
Thank you. We will look into it.

Also, whether we will be able to use custom-gst in the openwebrtc application for using this compressed video input?

Thanks.

**Comment by suganthikarthick** [comment]
Hi,
I downloaded the openwebrtc compressed branch and compiled and tested with test-send-receive application. My camera is supporting H264 compressed input, but still I am not able to see any video when executing the test-send-receive application.

Which corresponding version of Gstreamer I have to use or you have any comments on this?

**Comment by suganthikarthick** [comment]
Hi Ford Perfect,

We tested with opus encoded audio and that is working fine, but in case of H264 video, we are still facing some issues.

Below is the gst  debug messages that I got and the error information.

I have applied the pat hed 
0:00:26.259651822 ^[[331m12360^[[00m       0xddfa00 ^[[37mDEBUG  ^[[00m ^[[00;01;34m            GST_CAPS gstpad.c:2124:gst_pad_link_check_compatible_unlocked:<bin3:src>^[[00m src caps video/x-h264, stream-format=(string)avc, alignment=(string)au, profile=(string)baseline

0:00:26.259871706 ^[[331m12360^[[00m       0xddfa00 ^[[37mDEBUG  ^[[00m ^[[00;01;34m            GST_CAPS gstpad.c:2126:gst_pad_link_check_compatible_unlocked:<transport_bin_1:video_sink_0_2>^[[00m sink caps video/x-raw, format=(string)I420, width=(int)[ 1, 32767 ], height=(int)[ 1, 32767 ], framerate=(fraction)[ 0/1, 2147483647/1 ]; video/x-raw, width=(int)[ 1, 32767 ], height=(int)[ 1, 32767 ], framerate=(fraction)[ 0/1, 2147483647/1 ], format=(string){ I420, YV12, YUY2, UYVY, AYUV, RGBx, BGRx, xRGB, xBGR, RGBA, BGRA, ARGB, ABGR, RGB, BGR, YVYU, Y444, NV12, NV21, GRAY8, GRAY16_BE, GRAY16_LE }

0:00:26.260045824 ^[[331m12360^[[00m       0xddfa00 ^[[37mDEBUG  ^[[00m ^[[00;01;34m GST_CAPS gstpad.c:2144:gst_pad_link_check_compatible_unlocked:^[[00m caps are not compatible
0:00:26.260735113 ^[[331m12360^[[00m       0xddfa00 ^[[31;01mERROR  ^[[00m ^[[00m   owrtransportagent owr_transport_agent.c:780:handle_new_send_source:^[[00m Failed to link "(null)" with transport bin

Your suggestions will be valuable for us.

Thanks.

**Comment by ford-prefect** [comment]
I don't see video/x-h264 in the second line. What are those the caps of? Are you able to use your camera for h.264 using a simple gst-launch pipeline? Like ```gst-launch-1.0 <yourcamerasrc> ! h264parse ! decodebin ! autovideosink```

**Comment by suganthikarthick** [comment]
Yes, the camera is able to give x-h264

**Comment by suganthikarthick** [comment]
Hi,

As of now, we commented the flip related code in the owr_transport_agent.c, and now it is working with direct H264. This we checked with 1920x1080 resolution.

But, we have another camera with support 1440x1440 and it will only give H264 video source. With that camera, it is not working. 

We are getting the following errors.
0:00:13.865719750 ^[[332m 4315^[[00m 0x7f40f8003f20 ^[[37mDEBUG  ^[[00m ^[[00;01;34m            GST_CAPS gstpad.c:5251:pre_eventfunc_check:<capsfilter2:sink>^[[00m caps video/x-h264, stream-format=(string)avc, alignment=(string)au, width=(int)1440, height=(int)1440, pixel-aspect-ratio=(fraction)1/1, framerate=(fraction)30/1, profile=(string)high, parsed=(boolean)true, level=(string)4.1, codec_data=(buffer)01640029ffe1000c67640029ac1b4805a02d680201000468ee3cb0 not accepted

What could be the issue?

**Comment by philn** [comment]
This isn't the right forum for support requests/questions. Can you please use the owr google group for this? Thanks.

**Comment by superdump** [comment]
I would actually prefer use of individual issues on GitHub for each separate support issue instead of using the Google Group.

**Comment by suganthikarthick** [comment]
Hi,

As of now, we commented the flip related code in the owr_transport_agent.c, and now it is working with direct H264. This we checked with 1920x1080 resolution.

But, we have another camera with support 1440x1440 and it will only give H264 video source. With that camera, it is not working. 

We are getting the following errors.
0:00:13.865719750 ^[[332m 4315^[[00m 0x7f40f8003f20 ^[[37mDEBUG  ^[[00m ^[[00;01;34m            GST_CAPS gstpad.c:5251:pre_eventfunc_check:<capsfilter2:sink>^[[00m caps video/x-h264, stream-format=(string)avc, alignment=(string)au, width=(int)1440, height=(int)1440, pixel-aspect-ratio=(fraction)1/1, framerate=(fraction)30/1, profile=(string)high, parsed=(boolean)true, level=(string)4.1, codec_data=(buffer)01640029ffe1000c67640029ac1b4805a02d680201000468ee3cb0 not accepted

What could be the issue?
